### ISSUE TITLE
exp: simplify refs

### ISSUE THREAD

**Issue by dberenbaum** [iterative/dvc#9361]
Instead of refs like `exps/2f/52e2b57de94e8c6f302f6b1ad4b2863076aa92/sewed-gust`, it would be nicer to have exp refs like `exps/sewed-gust`.

Why?
1. Better to have global experiment names (shouldn't allow `sewed-gust` under two different hashes).
2. Easier for people to apply git operations to shorter refs. We can suggest people do operations directly on these refs, like `git branch sewed-gust-branch exps/sewed-gust` to make a branch (we could drop `dvc exp branch`). 
3. We can show only the exp name instead of showing both the exp name and sha everywhere (we do it now since it's hard to do git operations without the sha).

Since the current format includes the parent commit sha in the experiment path, I assume this would make it harder to connect experiments to their parent commits. What would be the implications of this change?

**Comment by daavoo** [comment]
> Also, where were the words taken from for the experiment names? cc @daavoo

Some random sampling of `adjectives` of length 4 + `nouns` of length 4 from an online wordlist on GitHub.
The hard coded lengths were to match the previous length of `exp-{hash}`.

> It would be nice to find better sets of words.

Yep. We could do a clean, manual collection of adjectives and nouns. 
It would be also easier to fill if we are more flexible with the resulting length.

**Comment by pmrowla** [comment]
> Since the current format includes the parent commit sha in the experiment path, I assume this would make it harder to connect experiments to their parent commits. What would be the implications of this change?

We rely on the part of the ref that contains parent commit sha to keep track of what experiments derive from a specific ref, and to make those lookups much easier (and faster). Making this change will have a significant negative performance hit on basically any experiments command that will scale with the total number of experiments a user has in their git repo

**Comment by pmrowla** [comment]
> @pmrowla Can we cache that info somewhere besides inside the exp refs themselves?

Yes, but if we are going to expect users to work with the refs directly, then users can make changes that would invalidate whatever we have already cached. Users can push/fetch/remove/modify the refs themselves with regular git commands and we would not know about it because it was not done through a DVC command. Exp refs as they are currently implemented, are not intended to be accessed directly by users.

This also gets way more complicated if we allow experiments to be a branch containing more than a single commit (i.e. checkpoints), since the baseline commit is not just the immediate parent of the tip exp ref commit. (This is another reason we have the current ref structure)

---

@dberenbaum is this seen as a high priority on the product side? I haven't noticed a significant number of user requests asking for this

**Comment by pmrowla** [comment]
One other thing to keep in mind is that we consider the entire `refs/exps/` namespace to be reserved by DVC, and we use a lot of things inside that namespace for things that are not named experiments. Making this change will also increase the likelihood that a user experiment name ends up conflicting with a "reserved" part of the namespace and ends up breaking DVC functionality (i.e. any experiment name that starts with `exec` will end up being completely hidden in `exp show` and other DVC commands)

> 2. Easier for people to apply git operations to shorter refs. We can suggest people do operations directly on these refs, like `git branch sewed-gust-branch exps/sewed-gust` to make a branch (we could drop dvc exp branch).

Users will still have to prefix everything in git commands with `refs/exps/...` (and not just `exps`), and using a direct ref does actually have behavior side effects in certain git commands (which are tuned to do specific things when given a raw `ref/...` name as opposed to a git branch, git tag, or SHA)

---

If the end goal is to just allow users to use exp names in git commands I would prefer to just add DVC command that takes any name/ref/sha/etc accepted by DVC commands and outputs the resolved SHA, so users can just do
```
$ git branch my-branch `dvc rev sewed-gust`
```

This is slightly more verbose than just `git branch my-branch refs/exps/sewed-gust`, but it still lets us keep control over the entire exp ref namespace and means the user doesn't have to use the entire refname

On the engineering side this is also something that can be implemented in 10 minutes (it's just a UI wrapper for `scm.resolve_rev()`) instead of requiring significant changes in the experiments implementation

**Comment by dberenbaum** [comment]
> @dberenbaum is this seen as a high priority on the product side? I haven't noticed a significant number of user requests asking for this

Not right now, just opening the issue for discussion.

> If the end goal is to just allow users to use exp names in git commands

Yeah, this is the goal.

> dvc rev sewed-gust

Sounds like a reasonable starting point. I don't think even this needs to be a p1 yet though.

**Comment by dberenbaum** [comment]
I don't think having a `dvc rev` command is much better than showing the sha, and don't see prioritizing this anytime soon, so closing
